packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -113,51 +110,84 @@ const withInfiniteScroll = (
         showSeeMore: false,
       };
 
-      this.onItemsFetch = props.onItemsFetch;
+      this.props.dataSource.onDataChange(this.updateData);
       this.observer = new IntersectionObserver(this.handleIntersection, {
         threshold: 1,
       });
     }
 
     componentDidMount() {
-      if (this.sentinel) {
-        this.observer.observe(this.sentinel);
+      this.loadMore().then(newState => {
+        this.setState(newState);
+      });
+    }
+
+    componentWillReceiveProps(newProps) {
+      if (this.props.dataSource !== newProps.dataSource) {
+        newProps.dataSource.onDataChange(this.updateData);
+        this.props.dataSource.offDataChange(this.updateData);
+        this.setState({
+          index: 0,
+          elementsToRender: [],
+          isListFinished: false,
+          showSeeMore: false,
+        });
       }
     }
 
-    componentDidUpdate() {
+    componentDidUpdate(_, prevState) {
       if (this.sentinel && this.state.index > 0) {
         this.observer.observe(this.sentinel);
       }
+      if (prevState.index !== 0 && this.state.index === 0) {
+        this.loadMore().then(newState => {
+          this.setState(newState);
```

Comment:
Since we are calling `setState` here anyway I think we should move all the `componentWillReceiveProps` here, we can change load more to accept the `index` and `elementsPerScroll` as argument.
```
if (prevState.dataSource !== this.state.dataSource) {
   prevState.dataSource.offDataChange(this.updateData);
   this.props.dataSource.onDataChange(this.updateData);
   this.loadMore(this.state.index, this.state.elementsPerScroll).then(newState => {
          this.setState({ ...newState, isListFinished: false });
   });
}

Reply:
yeah, that was another option. Changing it 👍

Reply:
change it. Working on the docs now